libkbfs/tlf_journal.go

```diff
@@ -1391,6 +1391,99 @@ func (j *tlfJournal) clearMDs(ctx context.Context, bid BranchID) error {
 	return j.mdJournal.clear(ctx, bid)
 }
 
+func (j *tlfJournal) doResolveBranch(ctx context.Context,
+	bid BranchID, blocksToDelete []BlockID, rmd *RootMetadata,
+	extra ExtraMetadata, mdInfo unflushedPathMDInfo,
+	perRevMap unflushedPathsPerRevMap) (mdID MdID, retry bool, err error) {
+	j.journalLock.Lock()
+	defer j.journalLock.Unlock()
+	if err := j.checkEnabledLocked(); err != nil {
+		return MdID{}, false, err
+	}
+
+	// The set of unflushed paths could change as part of the
+	// resolution, and the revision numbers definitely change.
+	if !j.unflushedPaths.reinitializeWithResolution(mdInfo, perRevMap) {
+		return MdID{}, true, nil
+	}
+
+	// First write the resolution to a new branch, and swap it with
+	// the existing branch, then clear the existing branch.
+	mdID, err = j.mdJournal.resolveAndClear(
+		ctx, j.config.Crypto(), j.config.encryptionKeyGetter(),
+		j.config.BlockSplitter(), bid, rmd)
+	if err != nil {
+		return MdID{}, false, err
+	}
+
+	// Then go through and mark blocks and md rev markers for ignoring.
+	err = j.blockJournal.ignoreBlocksAndMDRevMarkers(ctx, blocksToDelete)
+	if err != nil {
+		return MdID{}, false, err
+	}
+
+	// Finally, append a new, non-ignored md rev marker for the new revision.
+	err = j.blockJournal.markMDRevision(ctx, rmd.Revision())
+	if err != nil {
+		return MdID{}, false, err
+	}
+
+	j.signalWork()
+
+	// TODO: kick off a background goroutine that deletes ignored
+	// block data files before the flush gets to them.
+
+	return mdID, false, nil
+}
+
+func (j *tlfJournal) resolveBranch(ctx context.Context,
+	bid BranchID, blocksToDelete []BlockID, rmd *RootMetadata,
+	extra ExtraMetadata) (MdID, error) {
+	// Prepare the paths without holding the lock, as it might need to
+	// take the lock.  Note that the timestamp don't matter for the
+	// unflushed path cache.  This is a no-op if the unflushed path
+	// cache is uninitialized.  TODO: avoid doing this if we can
+	// somehow be sure the cache won't be initialized by the time we
+	// finish this put.
+	mdInfo := unflushedPathMDInfo{
+		revision: rmd.Revision(),
+		kmd:      rmd,
+		pmd:      *rmd.Data(),
+		// TODO: Plumb through clock?
+		localTimestamp: time.Now(),
+	}
+	perRevMap, err := j.unflushedPaths.prepUnflushedPaths(
```

Comment:
Since this is the second time we're doing this, and there's a fair bit of code duplication, might be worth defining a helper function for this that takes a closure that returns `retry, err`, and calls `prepUnflushedPaths` before calling it, and retries if necessary.